ci-bench/src/main.rs

```diff
@@ -0,0 +1,267 @@
+use std::io::{Read, Write};
+use std::ops::Deref;
+use std::sync::Arc;
+
+use clap::Parser;
+use rustls::{ClientConnection, ServerConnection};
+
+use crate::bench_lib::{black_box, Benchmark, BenchmarkCollection};
+use crate::cli::{Cli, Command};
+use crate::util::config::{new_client, new_server, BenchmarkParam, ClientAuth, ResumptionParam};
+use crate::util::key::KeyType;
+
+mod bench_lib;
+mod cli;
+mod rustls_facade;
+mod util;
+
+fn main() -> anyhow::Result<()> {
+    let benchmarks = get_benchmarks()?;
+
+    let cli = Cli::parse();
+    match cli.command {
+        Command::RunAll => {
+            let executable = std::env::args().next().unwrap();
+            let results = bench_lib::run_all(executable, &benchmarks)?;
+
+            // Output results in CSV (note: not using a library here to avoid extra dependencies)
+            for (name, instr_count) in results {
+                println!("{name},{instr_count}");
+            }
+        }
+        Command::RunSingle { index, side } => {
+            bench_lib::run_single(index, side, &benchmarks)?;
+        }
+        Command::Compare {
+            baseline_input,
+            candidate_input,
+        } => {
+            bench_lib::compare(baseline_input, candidate_input)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Returns all benchmarks wrapped in a [`BenchmarkCollection`]
+fn get_benchmarks() -> anyhow::Result<BenchmarkCollection> {
+    let mut benchmarks = Vec::new();
+    for param in all_benchmark_params() {
+        add_benchmark_group(&mut benchmarks, param);
+    }
+
+    BenchmarkCollection::new(benchmarks)
+}
+
+/// Returns the benchmark params to use for each group of benchmarks
+fn all_benchmark_params() -> Vec<BenchmarkParam> {
```

Comment:
Looks like this could be a `const` slice `&[BenchmarkParam]` instead, and we could ditch the trivial `BenchmarkParam::new()` method?

Reply:
Great idea! `const` doesn't work here, but `static ALL_BENCHMARK_PARAMS: &'static [BenchmarkParam]` does